libraries/Microsoft.Bot.Builder.Azure/AzureBlobStorage.cs

```diff
@@ -0,0 +1,154 @@
+        /// <returns></returns>
+        public async Task Delete(string[] keys)
+        {
+            if (keys == null) throw new ArgumentNullException(nameof(keys));
+
+            foreach (var blobName in keys.Select(GetBlobName))
+            {
+                var blobReference = this.Container.Value.GetBlobReference(blobName);
+                await blobReference.DeleteAsync(
+                    DeleteSnapshotsOption.None, new AccessCondition { IfMatchETag = "*" }, new BlobRequestOptions(), new OperationContext()).ConfigureAwait(false);
+            }
+        }
+
+        /// <summary>
+        /// Retrieve entities from the configured blob container
+        /// </summary>
+        /// <param name="keys">An array of entitiy keys</param>
+        /// <returns></returns>
+        public async Task<StoreItems> Read(string[] keys)
+        {
+            if (keys == null) throw new ArgumentNullException(nameof(keys));
+
+            var storeItems = new StoreItems();
+            foreach (string blobName in keys.Select(GetBlobName))
+            {
+                using (var memoryStream = new MemoryStream())
+                {
+                    var blobReference = this.Container.Value.GetBlobReference(blobName);
+                    await blobReference.DownloadToStreamAsync(memoryStream).ConfigureAwait(false);
+                    using (var streamReader = new StreamReader(memoryStream, Encoding.UTF8))
+                    {
+                        var json = streamReader.ReadToEnd();
+                        var obj = JsonConvert.DeserializeObject(json, serializationSettings);
+                        IStoreItem storeItem = obj as IStoreItem;
+                        if (storeItem != null)
+                        {
+                            storeItem.eTag = blobReference.Properties.ETag;
+                        }
+
+                        storeItems[blobName] = storeItem;
+                    }
+                }
+            }
+
+            return storeItems;
+        }
+
+        /// <summary>
+        /// Stores a new entity in the configured blob container
+        /// </summary>
+        /// <param name="changes"></param>
+        /// <returns></returns>
+        public async Task Write(StoreItems changes)
+        {
+            if (changes == null) throw new ArgumentNullException(nameof(changes));
+
+            foreach (var change in changes)
+            {
+                var blobName = GetBlobName(change.Key);
+                var newValue = change.Value as IStoreItem;
+                var json = JsonConvert.SerializeObject(newValue, Formatting.None, serializationSettings);
+                var blobReference = this.Container.Value.GetBlockBlobReference(blobName);
+
+                await blobReference.UploadTextAsync(json, new AccessCondition { IfMatchETag = newValue.eTag ?? "*" }, new BlobRequestOptions(), new OperationContext()).ConfigureAwait(false);

```

Comment:
Rather than serializing in memory and then uploading the string, you could technically serialize to a `JsonWriter` layered over a `StreamWriter` layered over the raw stream from the blob that you get from `CloudBlockBlob::OpenWriteAsync`. The gain here is, once again, less garbage being created, thus lowering GC overhead.